### ISSUE TITLE
Green image on 1660 Ti even with full precision

### ISSUE THREAD

**Issue by ArDiouscuros** [invoke-ai/InvokeAI#526]
**Describe your environment**
- GPU: 1660 Ti
- VRAM: 6 GB
- CPU arch: x86
- OS: Windows
- Python: miniconda
- Branch: main

**Describe the bug**
Image is flat green even with --full_precision

>> Loading model from models/ldm/stable-diffusion-v1/model.ckpt
LatentDiffusion: Running in eps-prediction mode
DiffusionWrapper has 859.52 M params.
making attention of type 'vanilla' with 512 in_channels
Working with z of shape (1, 4, 32, 32) = 4096 dimensions.
making attention of type 'vanilla' with 512 in_channels
>> Using slower but more accurate full-precision math (--full_precision)
>> Model loaded in 30.19s
>> Max VRAM used to load the model: 4.28G
>> Current VRAM usage:4.28G
>> Setting Sampler to k_lms

**Expected behavior**
Image is not green

**Comment by lstein** [comment]
This is a recurring problem involving use of mixed precision, and I'm sorry it's come back.

We'll get this figured out. In the meantime is there another fork of Stable Diffusion that works reliably in your hands?

**Comment by Pheenx** [comment]
> This is a recurring problem involving use of mixed precision, and I'm sorry it's come back.
> 
> We'll get this figured out. In the meantime is there another fork of Stable Diffusion that works reliably in your hands?

i have the 1660 super and i can only run the version from basujindal that actualy can create images that are not green in full precision mode hope that helps

**Comment by lstein** [comment]
That's a good data point. Thanks!

**Comment by mh-dm** [comment]
The problem here is there's no good way to configure both autocast and precision correctly at the same time. Currently autocast is picked only depending on device type (cuda/cpu/mps) in ldm/dream/devices.py. Precision is only "--full-precision" (i.e. the default torch.float32) or nothing.
@Pheenx and @ArDiouscuros try checking out the latest development branch, opening ldm/dream/devices.py and changing `return device_type,autocast` to `return device_type,nullcontext`.

Finally, I propose removing the full-precision flag and replace it with one "precision" flag with following options and behavior:
-default/unspecified: automatically pick float32/float16 as per device type (cuda/cpu/mps)
-float32: no autocast, no model.half()
-autocast: mixed precision allowed, no model.half(), as per the intentions of https://pytorch.org/docs/stable/amp.html
-float16: no autocast, model.to(torch.float16)
-bfloat16: no autocast, model.to(torch.bfloat16)

**Comment by Pheenx** [comment]
> The problem here is there's no good way to configure both autocast and precision correctly at the same time. Currently autocast is picked only depending on device type (cuda/cpu/mps) in ldm/dream/devices.py. Precision is only "--full-precision" (i.e. the default torch.float32) or nothing. @Pheenx and @ArDiouscuros try checking out the latest `development` branch, opening `ldm/dream/devices.py` and changing `return device_type,autocast` to `return device_type,nullcontext`.
> 
> Finally, I propose removing the full-precision flag and replace it with one "precision" flag with following options and behavior: -`default`/unspecified: automatically pick float32/float16 as per device type (cuda/cpu/mps) -`float32`: no autocast, no `model.half()` call -`autocast`: mixed precision allowed, no `model.half()` call (trying to match the intended usecase of https://pytorch.org/docs/stable/amp.html here) -`float16`: no autocast, `model.to(torch.float16)` -`bfloat16`: no autocast, `model.to(torch.bfloat16)`

omg thank you so much!!! That actualy worked with the --full_precision tag im so happy thank you

**Comment by lstein** [comment]
Try the latest development branch. I believe that the embiggen support wasn't fully merged, but just tested it and it seems fine now.

**Comment by lstein** [comment]
> The problem here is there's no good way to configure both autocast and precision correctly at the same time. Currently autocast is picked only depending on device type (cuda/cpu/mps) in ldm/dream/devices.py. Precision is only "--full-precision" (i.e. the default torch.float32) or nothing. @Pheenx and @ArDiouscuros try checking out the latest `development` branch, opening `ldm/dream/devices.py` and changing `return device_type,autocast` to `return device_type,nullcontext`.
> 
> Finally, I propose removing the full-precision flag and replace it with one "precision" flag with following options and behavior: -`default`/unspecified: automatically pick float32/float16 as per device type (cuda/cpu/mps) -`float32`: no autocast, no `model.half()` call -`autocast`: mixed precision allowed, no `model.half()` call (trying to match the intended usecase of https://pytorch.org/docs/stable/amp.html here) -`float16`: no autocast, `model.to(torch.float16)` -`bfloat16`: no autocast, `model.to(torch.bfloat16)`

This proposal looks very good to me. Can we set the precision and context in just one place where the model loads, or will the casing logic need to be scattered around various modules?

By the way, earlier this evening I tried to generate some images on a CPU-only Intel Xeon HPC node using --full_precision. It was slow compared to CUDA (8s/iteration), but it did work. Nice job on catching and fixing all the places where the precision was messed up. Half precision, by the way, gave me a runtime error of `"LayerNormKernelImpl" not implemented for 'Half'`, which seems like a reasonable error message.

**Comment by ArDiouscuros** [comment]
> This is a recurring problem involving use of mixed precision, and I'm sorry it's come back.
> 
> We'll get this figured out. In the meantime is there another fork of Stable Diffusion that works reliably in your hands?

Yes, I am using basujindal optimized repo, which with the latest changes allows me to generate 512x512 images in full precision mode, and it is also very quick in my opinion. I can literally hear that my card is working hard,  by the increased fan noise. =)

**Comment by ArDiouscuros** [comment]
> The problem here is there's no good way to configure both autocast and precision correctly at the same time. Currently autocast is picked only depending on device type (cuda/cpu/mps) in ldm/dream/devices.py. Precision is only "--full-precision" (i.e. the default torch.float32) or nothing.
> @Pheenx and @ArDiouscuros try checking out the latest `development` branch, opening `ldm/dream/devices.py` and changing `return device_type,autocast` to `return device_type,nullcontext`.
Thank you, change allowed me to generate image up to 384x384 and it is not green!

**Comment by lstein** [comment]
> The problem here is there's no good way to configure both autocast and precision correctly at the same time. Currently autocast is picked only depending on device type (cuda/cpu/mps) in ldm/dream/devices.py. Precision is only "--full-precision" (i.e. the default torch.float32) or nothing. @Pheenx and @ArDiouscuros try checking out the latest `development` branch, opening `ldm/dream/devices.py` and changing `return device_type,autocast` to `return device_type,nullcontext`.
> 
> Finally, I propose removing the full-precision flag and replace it with one "precision" flag with following options and behavior: -`default`/unspecified: automatically pick float32/float16 as per device type (cuda/cpu/mps) -`float32`: no autocast, no `model.half()` call -`autocast`: mixed precision allowed, no `model.half()` call (trying to match the intended usecase of https://pytorch.org/docs/stable/amp.html here) -`float16`: no autocast, `model.to(torch.float16)` -`bfloat16`: no autocast, `model.to(torch.bfloat16)`

@mh-dm, just to confirm, are you working on implementing this solution?

**Comment by mh-dm** [comment]
I haven't started yet (I will make a note here if I do). I'm looking into #508 first esp. since I think the training flow may be relevant to what precision options we should have.

**Comment by ArDiouscuros** [comment]
Could it be related to the that issue? 
[https://github.com/pytorch/pytorch/issues/77955](https://github.com/pytorch/pytorch/issues/77955)

**Comment by mh-dm** [comment]
I have started on switching from `full_precision` to `precision`

**Comment by mh-dm** [comment]
@wzgdesigntech on GTX 1650 is experiencing the same issue. However, since they have 4GB they can't use full precision. But more importantly @wzgdesigntech demonstrated that it's possible to use a float16 model and still get good output. The slow but working version of stable-diffusion used was:
https://github.com/basujindal/stable-diffusion/blob/main/optimizedSD/optimized_txt2img.py

This means that there's a change somewhere in lstein/stable-diffusion in that results in green images. Likely candidates are changes to
https://github.com/lstein/stable-diffusion/blob/development/ldm/modules/diffusionmodules/model.py
https://github.com/lstein/stable-diffusion/blob/development/ldm/modules/attention.py

@ArDiouscuros Pulling you into this as well since if the issue is found and you can use a float16 model then you'll very like be able to generate larger than 384x384 images. Btw, what's the max size images you can currently generate in the development branch?

As for next steps I'll soon post a branch with some changes where you can test if you still get green images.

**Comment by mh-dm** [comment]
I've created a new github branch for trying out potential fixes / tracking down the cause. You can obtain it through:

```sh
git fetch origin green-images-fix:green-images-fix
git checkout green-images-fix
```

To test out:
```sh
scripts/dream.py --precision=float16
[...initialization text..]
dream> brown corgi, clean background, flat design -s20 -Addim -S1
```
You should get an image like this:
![000010 1](https://user-images.githubusercontent.com/299015/191012247-6efc79ca-e942-41d9-ab8a-180973749767.png)

@ArDiouscuros since you have more VRAM you can also try full precision/float32 but with autocast enabled:
```sh
scripts/dream.py --precision=autocast
```

**Comment by lstein** [comment]
@mh-dm Thanks for taking the lead on this!

**Comment by ArDiouscuros** [comment]
> @wzgdesigntech on GTX 1650 is experiencing the same issue. However, since they have 4GB they can't use full precision. But more importantly @wzgdesigntech demonstrated in discussion #584 that it's possible to use a float16 model and still get good output. The slow but working version of stable-diffusion used was: https://github.com/basujindal/stable-diffusion/blob/main/optimizedSD/optimized_txt2img.py

@mh-dm 

I have my doubts. I am using the same repo basujindal and optimized_txt2img.py
If I call it like this: 
python optimizedSD/optimized_txt2img.py --prompt "brown corgi, clean background, flat design" --H 512 --W 512 --seed 42 --n_iter 1 --n_samples 1 --ddim_steps 20 --precision full --turbo --unet_bs 1
I will get a picture:
![image](https://user-images.githubusercontent.com/72071512/191076548-a63f6ba9-2258-48a4-b2e2-3eae78789048.png)

If I omit full precision:
![image](https://user-images.githubusercontent.com/72071512/191076731-5d256152-9b6e-4bfd-9679-de7819606dde.png)

> @ArDiouscuros Pulling you into this as well since if the issue is found and you can use a float16 model then you'll very like be able to generate larger than 384x384 images. Btw, what's the max size images you can currently generate in the development branch? (asking cause there's been recent optimizations)?
Was not able to go higher than 384, if it is 64 pixels more I get exception after the sampling is complete. 
 
> As for next steps I'll soon post a branch with some changes where you can test if you still get green images.
I am interested very much to test it...

**Comment by mh-dm** [comment]
Have you tried the `green-images-fix` branch? My last post had instructions.

**Comment by mh-dm** [comment]
Also now I notice that @wzgdesigntech was passing `--precision full` to `optimized_txt2img.py`. https://github.com/lstein/stable-diffusion/discussions/584#discussioncomment-3678914
Finally took a closer look into the script. I thought the way it still worked on 4GB was that they called model.half() but it turns out they split the model and computation into parts so only part of the model has to be in VRAM at a time. That's a much more involved change to do esp. since we don't want to slow anyone else down.

I also took a quick at pytorch/pytorch#77955 . Could be related but doesn't really give me any further leads.

**Comment by ArDiouscuros** [comment]
> ```shell
> scripts/dream.py --precision=float16
> ```

Could not get it working...
(ldm) G:\Programs\SDAI\StableDiffusionManual\repo\lstein\stable-diffusion>python scripts/dream.py --precision=float16
* Initializing, be patient...

Traceback (most recent call last):
  File "G:\Programs\SDAI\StableDiffusionManual\conda\envs\ldm\lib\site-packages\skimage\__init__.py", line 151, in <module>
    from ._shared import geometry
ImportError: cannot import name 'geometry' from 'skimage._shared' (G:\Programs\SDAI\StableDiffusionManual\conda\envs\ldm\lib\site-packages\skimage\_shared\__init__.py)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "G:\Programs\SDAI\StableDiffusionManual\repo\lstein\stable-diffusion\scripts\dream.py", line 356, in <module>
    main()
  File "G:\Programs\SDAI\StableDiffusionManual\repo\lstein\stable-diffusion\scripts\dream.py", line 38, in main
    from ldm.generate import Generate
  File "g:\programs\sdai\stablediffusionmanual\repo\lstein\stable-diffusion\ldm\generate.py", line 19, in <module>
    import skimage
  File "G:\Programs\SDAI\StableDiffusionManual\conda\envs\ldm\lib\site-packages\skimage\__init__.py", line 154, in <module>
    _raise_build_error(e)
  File "G:\Programs\SDAI\StableDiffusionManual\conda\envs\ldm\lib\site-packages\skimage\__init__.py", line 131, in _raise_build_error
    raise ImportError("""%s
ImportError: cannot import name 'geometry' from 'skimage._shared' (G:\Programs\SDAI\StableDiffusionManual\conda\envs\ldm
\lib\site-packages\skimage\_shared\__init__.py)
It seems that scikit-image has not been built correctly.

Your install of scikit-image appears to be broken.
Try re-installing the package following the instructions at:
https://scikit-image.org/docs/stable/install.html

**Comment by mh-dm** [comment]
That's from a new dependency for development which you can satisfy with `pip install scikit-image`

**Comment by ArDiouscuros** [comment]
@mh-dm 
> I've created a new github branch for trying out potential fixes / tracking down the cause.

Testing new branch, I like how it works without any additional tuning. But lower precision modes still not work for me.
So if I just run python scripts\dream.py I am able to get correct image:
```
(ldm) G:\Programs\SDAI\StableDiffusionManual\repo\lstein\stable-diffusion>python scripts\dream.py
* Initializing, be patient...
Using more accurate float32 precision
>> Model loaded in 29.74s
>> Max VRAM used to load the model: 4.28G
>> Current VRAM usage:4.28G

dream> brown corgi, clean background, flat design -s20 -W512 -H512 -C7.5 -Addim -S1
>> Setting Sampler to ddim
>> Usage stats:
>>   1 image(s) generated in 57.73s
>>   Max VRAM used for this generation: 5.17G. Current VRAM utilization: 4.36G
>>   Max VRAM used since script start:  5.17G
```
![image](https://user-images.githubusercontent.com/72071512/191126196-a1aadee5-ed20-4479-9583-cc0acca98790.png)

```
(ldm) G:\Programs\SDAI\StableDiffusionManual\repo\lstein\stable-diffusion>python scripts/dream.py --precision=float16
* Initializing, be patient...
Using faster float16 precision
>> Model loaded in 20.29s
>> Max VRAM used to load the model: 2.17G
>> Current VRAM usage:2.17G
dream> brown corgi, clean background, flat design -s20 -W512 -H512 -C7.5 -Addim -S1
>> Usage stats:
>>   1 image(s) generated in 52.57s
>>   Max VRAM used for this generation: 3.35G. Current VRAM utilization: 2.21G
>>   Max VRAM used since script start:  3.35G
```
Result is total black 

```
(ldm) G:\Programs\SDAI\StableDiffusionManual\repo\lstein\stable-diffusion>python scripts/dream.py --precision=autocast
* Initializing, be patient...
Using more accurate float32 precision
>> Model loaded in 16.59s
>> Max VRAM used to load the model: 4.28G
>> Current VRAM usage:4.28G
dream> brown corgi, clean background, flat design -s20 -W512 -H512 -C7.5 -Addim -S1
>> Usage stats:
>>   1 image(s) generated in 60.59s
>>   Max VRAM used for this generation: 5.04G. Current VRAM utilization: 4.36G
>>   Max VRAM used since script start:  5.04G
```
Same black picture.

Comparing with other repo: it runs faster despite the fact it splits model into 3 parts:
```
(ldo) G:\Programs\SDAI\StableDiffusionManual\repo\basujindal\stable-diffusion>python optimizedSD/optimized_txt2img.py --
prompt "brown corgi, clean background, flat design" --H 512 --W 512 --seed 1 --n_iter 1 --n_samples 1 --ddim_steps 20 --
precision full --turbo --unet_bs 1
Running PLMS Sampling with 20 timesteps
memory_final =  3.628544
Samples finished in 0.65 minutes and exported to outputs/txt2img-samples\brown_corgi,_clean_background,_flat_design
```
But results are not identical:
![image](https://user-images.githubusercontent.com/72071512/191127087-633cdfb4-0662-44d1-98e0-d568791e800b.png)

**Comment by mh-dm** [comment]
For the last command, results are not identical because it looks like `optimized_txt2img.py` is using (defaulting to) `PLMS Sampling`. So it's fine/expected since the dream.py command I showed was with `-Addim`, i.e. using the DDIM sampler.

In terms of ways forward, here's something to try. Following the link you posted there's this very relevant and fixed issue in pytorch https://github.com/pytorch/pytorch/issues/58123 and the fix landed in cudnn8.2.2. However, if following the standard install you're likely to end up with cudnn8.2.0. You can verify the version you have with:

```python
>>> import torch
>>> torch.backends.cudnn.version()
8200
```

I would say to try updating pytorch but with a note it's possible to break things. Here's a command to try:

```sh
conda install pytorch cudatoolkit=11.3 -c pytorc
```

Make note of what versions it updates from so you can more easily undo. If updating doesn't help you can also try updating cudatoolkit. Potentially useful commands: `conda search pytorch::pytorch` and `conda search cudatoolkit` to show you what versions you can try but also note that the pytorch build (like `py3.9_cuda11.3_cudnn8.3.2_0`) must match your python and the cudatoolkit version). The default versions are here: https://github.com/invoke-ai/InvokeAI/blob/development/environment.yaml

**Comment by ArDiouscuros** [comment]
@mh-dm 
Thanks for the info about the ways forward, I have already started following:
1. Install updated version of pytorch with cuda toolkit 
```
pip install torch==1.12.0+cu116 torchvision==0.13.0+cu116 --extra-index-url https://download.pytorch.org/whl/cu116
```
Result is - image is not green, it is black, for all prompts and settings
2. Reset all to defaults and manually downloaded cudnn files and replace in ..\conda\envs\ldm\Lib\site-packages\torch\lib\, tested following versions
```
8.2.0.53
8.2.2.26
8.2.4.15
8.3.3.40
8.4.1.50
8.5.0.96
```
Libraries are from NVIDIA https://developer.nvidia.com/cudnn
With current versions:
```
# Name                   Version       Build  Channel
pytorch                  1.11.0          py3.8_cuda11.3_cudnn8_0    pytorch
cudatoolkit              11.3.1         h59b6b97_2
>>> import torch
>>> torch.backends.cudnn.version()
8500
```
The same black picture result
4. Downgraded torch to 1.10.1+cu102 - stable diffusion could not start

But as image is not green anymore, and NaN tests are passing if I replace cudnn to versions > 8.2 I feel that I am on correct way but something is still missing...
Any ideas?

**Comment by mh-dm** [comment]
You mentioned before that you're getting black output in the `green-images-fix` branch. Is the green-to-black change happening on latest development (where you can now use --precision) with just the cudnn update?

**Comment by ArDiouscuros** [comment]
> Is the green-to-black change happening on latest development (where you can now use --precision) with just the cudnn update?

Yes, I had run following test with cudnn 8200 and 8500:
development branch, float16, 8200, prompt test image -s1 -S1 => green image
development branch, float32, 8200, prompt test image -s1 -S1 => crash on save
development branch, float32, 8200, free gpu mem, prompt test image -s1 -S1 => normal image

development branch, float16, 8500, prompt test image -s1 -S1 => black image
development branch, float32, 8500, prompt test image -s1 -S1 => normal image
development branch, float32, 8500, free gpu mem, prompt test image -s1 -S1 => normal image

**Comment by mh-dm** [comment]
Interesting. Couple things to try before I'm out of ideas:

- `--precision=autocast` at development head
- `optimizedSD/optimized_txt2img.py` with the updated cudnn but without `--precision=full`

**Comment by ArDiouscuros** [comment]
--precision=autocast for me is the same as float32. It crashes 2 times out of 3 after sampling, 1 time produces normal image

with optimized repo I have got interesting results:
cudnn 8500, no full precision, 50 steps  => black image in 2.75 minutes
cudnn 8500, full precision, 50 steps       => normal image in 0.95 minutes
float32 is more then twice fast in optimized repo

I checked on development branch 
cudnn 8500, float32, 50 steps => crash  - 0 image(s) generated in 189.93s
cudnn 8500, float32, 50 steps => normal 1 image in 192.18s
cudnn 8500, float16, 50 steps => normal image in 127.70s

Soo... looks like 16xx cards are broken?
e.g. see the same perf issue here: https://github.com/tensorflow/tensorflow/issues/44191
https://stackoverflow.com/questions/69188051/pytorchs-nn-conv2d-with-half-precision-fp16-is-slower-than-fp32

I would really like to get optimized repo speed + invoke-ai features )

**Comment by mh-dm** [comment]
"cudnn 8500, float16, 50 steps => normal image in 127.70s"
Did you mean black image? (To match the results you reported before)

Do you get a stack trace for the crashes?

**Comment by ArDiouscuros** [comment]
> Did you mean black image? (To match the results you reported before)
Opps, copy paste error, it was black.

Stack trace
```
Traceback (most recent call last):
  File "\stable-diffusion\ldm\generate.py", line 411, in prompt2image
    results = generator.generate(
  File "\stable-diffusion\ldm\dream\generator\base.py", line 72, in generate
    image = make_image(x_T)
  File "\conda\envs\ldm\lib\site-packages\torch\autograd\grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "\stable-diffusion\ldm\dream\generator\txt2img.py", line 50, in make_image
    return self.sample_to_image(samples)
  File "\stable-diffusion\ldm\dream\generator\base.py", line 84, in sample_to_image
    x_samples = self.model.decode_first_stage(samples)
  File "\conda\envs\ldm\lib\site-packages\torch\autograd\grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "\stable-diffusion\ldm\models\diffusion\ddpm.py", line 1117, in decode_first_stage
    return self.first_stage_model.decode(z)
  File "\stable-diffusion\ldm\models\autoencoder.py", line 421, in decode
    dec = self.decoder(z)
  File "\conda\envs\ldm\lib\site-packages\torch\nn\modules\module.py", line 1110, in _call_impl
    return forward_call(*input, **kwargs)
  File "\stable-diffusion\ldm\modules\diffusionmodules\model.py", line 599, in forward
    h = self.up[i_level].block[i_block](h, temb)
  File "\conda\envs\ldm\lib\site-packages\torch\nn\modules\module.py", line 1110, in _call_impl
    return forward_call(*input, **kwargs)
  File "\stable-diffusion\ldm\modules\diffusionmodules\model.py", line 121, in forward
    h = silu(h)
  File "\conda\envs\ldm\lib\site-packages\torch\nn\functional.py", line 2032, in silu
    return torch._C._nn.silu(input)
RuntimeError: CUDA out of memory. Tried to allocate 256.00 MiB (GPU 0; 6.00 GiB total capacity; 4.57 GiB already allocated; 133.20 MiB free;
 4.71 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See
 documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
```

**Comment by ArDiouscuros** [comment]
Just have got some success in generating image on GTX 1660 Ti + float16 + cudnn 8.5
Standard installation of InvokeAI
Download cudnn libraries from NVIDIA site
Replace existing cudnn files in conda\envs\ldm\Lib\site-packages\torch\lib\
Add missing dependency zlibwapi.dll
**After that we get black images instead of green**

Add somewhere in init scripts (e.g. after model is loaded):
**torch.backends.cudnn.benchmark = True**

After that we get normal images in float16 mode.
But, performance of sampling in this mode is 2 - 3 times **slower** for GTX 16xx cards.

**Comment by jdries3** [comment]
Tonight I completely failed to get today's 2.0 release running on a 1650ti (has 4GB).

I tried using the AUTOMATIC1111 fork since it says it has support for 4GB and even 3GB VRAM GPUs. All of the default runs worked fine in terms of not generating errors, but the resulting images were all black. After looking more closely at how they implemented their `medvram` and `lowvram` support (https://github.com/AUTOMATIC1111/stable-diffusion-webui/blob/122d42687b97ec4df4c2a8c335d2de385cd1f1a1/modules/lowvram.py), I was able to find a set of command line parameters that worked for me:

```
--lowvram --precision full --no-half --always-batch-cond-uncond --unload-gfpgan --opt-split-attention
```

I haven't looked to see how much of what they do to manage models between GPU and CPU, but their approach might be useful to find a path forward.

**Comment by ArDiouscuros** [comment]
> Tonight I completely failed to get today's 2.0 release running on a 1650ti (has 4GB).

Could you be more specific please? Did you get OOM, other crash, or image was just green/black?

Did you try to use method described in previous message and to put updated cudnn libs to your installation?

**Comment by jdries3** [comment]
Yes. I got OOM errors. I don't have the device in front of me at the moment, but it would fail on allocating a 20MB segment, give me a report out of the VRAM used (typically 3.43GB of 4.0GB) and suggested that I set `PYTORCH_CUDA_ALLOC_CONF` to something like `max_split_size_mb:128` and `garbage_collection_threshold:0.6`. The `garbage_collection_threshold` option isn't available in pytorch 1.11, so I had to upgrade to 1.12.1, resolve the conflicts. I didn't try the optimized repo from `https://github.com/basujindal/stable-diffusion/`, but instead switched to the AUTOMATIC1111 fork.

I was contemplating trying to bring the work done in that fork back to the InvokeAI fork when I found this thread, and figured I would lend my experience back. I did (originally) get just black images from the AUTOMATIC1111 fork, until I altered the command line arguments around precision, lowvram, and the other flags. I didn't experiment with precision autocast or whether I could get away w/o unloading gfpgan or split-attention options as I had been working to get something working for a few hours.

If InvokeAI has a path forward, I won't bother to look more deeply or try to port over their work.
